File: ACM/Acwing/831.py
'''
Author: NEFU AB-IN
Date: 2022-02-27 21:04:00
FilePath: \ACM\Acwing\831.py
LastEditTime: 2022-02-27 21:14:38
'''
N = int(1e5 + 10)
h, p = [0] * N, [0] * N
p[0] = 1

P, MOD = 131, 1 << 64


def get(l, r):
    return (h[r] - h[l - 1] * p[r - l + 1] % MOD) % MOD


n_ss = int(input())
ss = input()
ss = " " + ss
res_ss = 0
for i in range(1, n_ss + 1):
    res_ss = (res_ss * P % MOD + ord(ss[i])) % MOD

n_s = int(input())
s = input()
s = " " + s

for i in range(1, n_s + 1):
    h[i] = (h[i - 1] * P % MOD + ord(s[i])) % MOD
    p[i] = p[i - 1] * P % MOD

for i in range(1, n_s + 1):
    if get(i, i + n_ss - 1) == res_ss:
        print(i - 1, end=" ")

File: ACM/Acwing/831/831.cpp
/*
 * @Author: NEFU AB-IN
 * @Date: 2023-02-23 23:58:49
 * @FilePath: \Acwing\831\831.cpp
 * @LastEditTime: 2023-02-24 00:03:22
 */
#include <bits/stdc++.h>
using namespace std;
#define int long long
#undef int

#define SZ(X) ((int)(X).size())
#define ALL(X) (X).begin(), (X).end()
#define IOS                                                                                                            \
    ios::sync_with_stdio(false);                                                                                       \
    cin.tie(nullptr);                                                                                                  \
    cout.tie(nullptr)
#define DEBUG(X) cout << #X << ": " << X << '\n'
typedef pair<int, int> PII;

const int N = 1e5 + 10, INF = 0x3f3f3f3f;

int ne[N];
signed main()
{
    IOS;
    int n, m;
    string p, s;
    cin >> n >> p >> m >> s;

    p = " " + p;
    s = " " + s;
    for (int i = 2, j = 0; i <= n; ++i)
    {
        while (j && p[i] != p[j + 1])
            j = ne[j];
        if (p[i] == p[j + 1])
            j++;
        ne[i] = j;
    }
    for (int i = 1, j = 0; i <= m; i++)
    {
        while (j && s[i] != p[j + 1])
            j = ne[j];
        if (s[i] == p[j + 1])
            j++;
        if (j == n)
        {
            cout << i - n << " ";
            j = ne[j];
        }
    }

    return 0;
}